http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/auth/SamAuthProvider.scala

```diff
@@ -127,6 +127,55 @@ class SamAuthProvider[F[_]: OpenTelemetryMetrics](
     } yield (callerActions, projectCallerActions)
   }
 
+  /**
+   * Get all Sam resource IDs of the given type on which the user is granted permissions. Returns explicitly granted
+   * policies only; users with workspace owner access, for example, can see all runtimes in the workspace, but
+   * getAuthorizedIds[SamRuntimeResourceId] is not guaranteed to return all runtimes in that workspace. Works for any
+   * Sam resource type except SharedApp and similar alternative app types; these require a specific Sam resource object
+   * to determine permissions, as these are stored on the individual records.
+   * @param isOwner whether to return only resources that the user is granted Owner permission on (or the equivalent for
+   *                the resource type, defined as `ownerRoleName` and `policyNames` on the SamResource)
+   * @param userInfo the current user
+   * @param samResource the resource archetype (App, Runtime, etc)
+   * @param decoder converts Json -> SamResourceId, needed by SamDao
+   * @param ev the request context
+   * @tparam R SamResourceId
+   * @return List[SamResourceId] of all resources of type R which are visible to the user at the given access level
+   */
+  override def getAuthorizedIds[R](
+    isOwner: Boolean,
+    userInfo: UserInfo
+  )(implicit
+    samResource: SamResource[R],
+    decoder: Decoder[R],
+    ev: Ask[F, TraceId]
+  ): F[List[R]] = {
+    val authHeader = Authorization(Credentials.Token(AuthScheme.Bearer, userInfo.accessToken.token))
+    val ownerRole: SamRole = samResource.ownerRoleName
+    val isOwnerInPolicies: Boolean = samResource.policyNames.exists { policyName: SamPolicyName =>
+      ownerRole.asString == policyName.toString
```

Comment:
It feels smelly to be doing to-string comparisons of two different objects, but as far as I can tell both SamRole and SamPolicyName originate in the same information parsed from Sam JSON responses.

Reply:
you can look at the underlying API that's being used, and compare that to https://sam.dsde-dev.broadinstitute.org/ to figure out if anything is deprecated..My guess is no

Reply:
I don't mean deprecated in Sam's API; sorry for my vagueness. I mean, which of these two values is the point of truth in Leo, and what are they used for? It seems like the SamAuthProvider is just using SamPolicyName.Owner directly when it wants to check that status, but for some resources the ownerRoleName is something else, like Creator. Is that field used somewhere I am missing?

Reply:
... did a quick scan of the code, should we be using something more along the lines of `#hasPermission()` to determine the ability to do X?

Reply:
@cpate4 I think you're suggesting a move in a good direction, but I'm not certain where to draw the boundary within this task specifically. If you have time let's talk about it!

Reply:
My initial answer is that this seems to *be* how we handle permission checking: we ask Sam to give us the ids and associated policy (role) names that the user is granted for a given resource type, then filter to any ids whose policies intersect with Leo's statically configured lists of visible policy names.

As far as I can tell, `hasPermission` requires the context of a specific resource ID. We don't have that at this stage of this method.

Reply:
https://broadworkbench.atlassian.net/browse/IA-4295 is an existing ticket which will partly address this concern: use Sam's client library instead of our arcane homunculus.

Created https://broadworkbench.atlassian.net/browse/IA-4611 to map the larger effort to get Leo out of the business of authentication.

https://broadworkbench.atlassian.net/browse/IA-2093 also seems related.

Reply:
@cpate4 here's where Sam configures the resource type `notebook-cluster`, which maps to v1 GCP runtimes: https://github.com/broadinstitute/sam/blob/cb1ca54e36da9bfedd9f375f366a143c232425b7/src/main/resources/reference.conf#L735C6-L735C6

And here's where Leo duplicates the definition (with the nonfunctional additional context of a specific resource ID, which does not affect the evaluated value of any of the fields here): https://github.com/DataBiosphere/leonardo/blob/1945671202b9ddf4cd9a11a780d0a324cb398f16/http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/model/LeoAuthProvider.scala#L34

The question I'm asking in calculating `canOwnerRead` is "given a resourceType (in general, not in a user context) is `ownerRoleName` in `policyNames`?"

The reason I'm doing a string comparison instead of object-to-object is because `SamRole` and `SamPolicyName` are separate classes (which I'm [pretty sure is technical debt](https://broadworkbench.atlassian.net/browse/IA-4608)).

Reply:
I've merged some work to make this clearer; this flow relies only on `SamRole` now. `SamPolicyName` should be deleted.